### ISSUE TITLE
must calculate indentation when user presses "enter"

### ISSUE THREAD

**Issue by oakmac** [oakmac/atom-parinfer#14]
We need to use Parinfer to calculate the proper indentation length when the user presses "enter" to go to the next line.

Essential feature.

**Comment by 00Davo** [comment]
Indentation-on-enter indeed doesn't seem to be working correctly, at least on my setup, which makes writing even simple code with Parinfer active awfully difficult. For instance:
```lisp
; | is my cursor
(defun id (x)|)
; enter
(defun id (x))
|
; manual indentation now required to start the function body!
```

The correct indentation does appear to be applied in this case, where the indent merely must match the previous line exactly:
```scheme
(define blahs '(a
                b 
                c|))
; enter
(define blahs '(a
                b 
                c))
                |
; cursor's in perfect spot to add extra item to the list, directly under c
```
But the very common case where the current form must be continued, by *increasing* indent on Enter, is not successfully supported. Note that the [official Parinfer demos](https://shaunlebron.github.io/parinfer/) produce correct indentation in both of these examples.

I've tested under both [language-lisp](https://atom.io/packages/language-lisp) and [language-racket](https://atom.io/packages/language-racket), with exactly the same results - I'm not sure whether the active filetype actually affects Atom's indentation logic anyway, but it didn't help here.

**Comment by neverfox** [comment]
Yeah, this is really disappointing. Has anybody come up with a fix (to what I assume is a problem with Atom's Clojure indentation logic)?

**Comment by oakmac** [comment]
I have been annoyed with this recently too.

**Comment by neverfox** [comment]
Does that imply that it wasn't always this way?

**Comment by oakmac** [comment]
I'm not sure.

**Comment by neverfox** [comment]
Another indent mode example for the record

```clojure
(def m {:a 1|})

(def m {:a 1})
  |
```
Six spaces off from where you need it to continue the map.

**Comment by shaunlebron** [comment]
Thanks for the examples @00Davo and @neverfox 

The crux of the problem is that Atom [does not indent based on language grammars](https://github.com/atom/atom/issues/9059).  The python community has been [most vocal](https://github.com/atom/language-python/issues/22) about this, and they recently created a [stopgap package](https://atom.io/packages/python-indent) for it.  We could follow by taking cues from indentation rules documented for [emacs](https://github.com/clojure-emacs/clojure-mode#indentation-options) or [vim](https://github.com/guns/vim-clojure-static#indent-options), or do a quick, _partial_ solution using parinfer itself:

_When enter is pressed:_

1. Set indentation of the current line to that of the previous line.
2. Run paren-mode (synchronously) at the cursor.
3. Resume indent-mode (debounced).

This will result in only a one-space indent for parens, which will be mildly annoying for the first indented line:

```clj
(def foo|)

(def foo)
 |
```

But the following lines will resume your two-space indentation if the previous line has it:

```clj
(def foo
  bar|)

(def foo
  bar)
  |
```

**Comment by shaunlebron** [comment]
I go this working when pressing enter inside a Paren Trail, but I have to think more about how to do this when pressing enter further inside a line.  Tracking at https://github.com/shaunlebron/parinfer/issues/97

**Comment by rgdelato** [comment]
In the new [Atom guide](https://clojurescript.org/tools/atom) on the ClojureScript site, I pointed people towards the solution on Jason Gilman's [opinionated setup](https://gist.github.com/jasongilman/d1f70507bed021b48625) page, which is to install the Paredit plugin and only assign the keybinding for the "Enter" key.

If anyone finds a better solution or if Parinfer ends up at a place where it can do this on it's own, please send a PR to the ClojureScript site!

**Comment by shaunlebron** [comment]
This is actually fixed in latest Parinfer:
https://runkit.com/shaunlebron/595d8d7fb05c230013f8dd60

dependent on https://github.com/oakmac/atom-parinfer/pull/78 being merged

**Comment by rgdelato** [comment]
@shaunlebron After playing with it, Parinfer's auto-indentation feels a bit weird. But I can continue that discussion in https://github.com/shaunlebron/parinfer/issues/59 rather than here.
